### PR TITLE
Allow using an existing sled::Db with SledStateStore

### PR DESCRIPTION
**Pull request by gferon**
This helps when you want to reuse an already opened `sled::Db` (to save on opened files or flushing threads, for example).

I'm more than happy to change the approach, and maybe add a `db` field to the `SledStateStoreBuilder` for example.

### CODE REVIEW

**Comment by Hywan** [comment]
Hello,

Thank you for your contributions.

I reckon it's preferable to use `SledStateStoreBuilder` to build a `SledStateStore` with an existing `Db` rather than making `open_helper` public. The builder pattern is design specifically to solve this kind of problem :-).

Thoughts?

**Comment by gferon** [comment]
Thanks for the feedback! I added a `db` field to the builder, but I'm not sure how to best handle the case where you provide both a `db` and a `path`.

**Comment by jplatte** [comment]
I think we should have them override each other. Take a look at how it works in `ClientBuilder`, where `homeserver_url` and `server_name` override each other.

**Comment by gferon** [comment]
> I think we should have them override each other. Take a look at how it works in `ClientBuilder`, where `homeserver_url` and `server_name` override each other.

If by overriding you mean setting `db` should take precedence over `path`, then that's the behaviour in this PR: https://github.com/matrix-org/matrix-rust-sdk/pull/1186/files#diff-dd89bc717be41cc9fd3a4a3d8c47d742e7afffd9a61d2fccff94b3f570665b22R198

**Comment by jplatte** [comment]
No, I mean whatever was called last should be what is used. As in the builder should have one field of an enum type that can hold *either* a `Db` or a path.

**Comment by gferon** [comment]
I see. It should be good now!

File: crates/matrix-sdk-sled/src/state_store.rs [hunk 1]
```diff
@@ -188,14 +192,23 @@ pub struct SledStateStoreBuilderConfig {
 
 impl SledStateStoreBuilder {
     pub fn path(&mut self, path: PathBuf) {
-        self.path = Some(path.join("matrix-sdk-state"));
+        self.db_or_path = Some(DbOrPath::Path(path.join("matrix-sdk-state")));
```
Comment (jplatte) [inline crates/matrix-sdk-sled/src/state_store.rs:195]: Can we do the path joining inside `build` as before please?

**Review by jplatte — COMMENTED** [review]
One nitpick, otherwise looks good.

File: crates/matrix-sdk-sled/src/state_store.rs (hunk 1 above)
Comment (gferon) [reply crates/matrix-sdk-sled/src/state_store.rs]: Sure, had to change the structure a little bit to not repeat myself.

File: crates/matrix-sdk-sled/src/state_store.rs [hunk 2]
```diff
@@ -170,11 +170,18 @@ const ALL_GLOBAL_KEYS: &[&str] = &[VERSION_KEY];
...
+    Db(Db),
+    Path(PathBuf),
+}
+
+#[derive(Builder, Debug)]
 #[builder(name = "SledStateStoreBuilder", build_fn(skip))]
 pub struct SledStateStoreBuilderConfig {
-    /// Path to the sled store files, created if not yet existing
-    path: PathBuf,
+    /// Use an existing open sled database
```
Comment (jplatte) [inline crates/matrix-sdk-sled/src/state_store.rs:182]: This is now wrong. I think you can remove it here entirely.

File: crates/matrix-sdk-sled/src/state_store.rs [hunk 3]
```diff
@@ -184,22 +191,28 @@ pub struct SledStateStoreBuilderConfig {
...
-        let mut cfg = Config::new().temporary(is_temp);
+    pub fn path(&mut self, path: PathBuf) {
+        self.db_or_path = Some(DbOrPath::Path(path));
+    }
 
-        let path = if let Some(path) = &self.path {
-            let path = path.join("matrix-sdk-state");
+    pub fn db(&mut self, db: Db) {
+        self.db_or_path = Some(DbOrPath::Db(db));
+    }
```
Comment (jplatte) [inline crates/matrix-sdk-sled/src/state_store.rs:200]: These functions should be documented.

File: crates/matrix-sdk-sled/src/state_store.rs [hunk 4]
```diff
@@ -184,22 +191,28 @@ pub struct SledStateStoreBuilderConfig {
...
 
-            cfg = cfg.path(&path);
-            Some(path)
-        } else {
-            None
+    pub fn build(&mut self) -> Result<SledStateStore> {
```
Comment (jplatte) [inline crates/matrix-sdk-sled/src/state_store.rs:202]: Same here really, but that's an existing issue so feel free to ignore.

**Review by jplatte — COMMENTED** [review]
Thanks, implementation looks good. I just noticed the docs still need work though.

**Comment by gferon** [comment]
I added the missing doc comments, and based the one from `SledStateStoreBuilder::build` on the style of `ClientBuilder`. Feel free to edit them directly if you prefer to word things differently! Hope this is good to go now, and thanks for all the very quick feedback 😄

File: crates/matrix-sdk-sled/src/state_store.rs [hunk 5]
```diff
@@ -184,22 +190,40 @@ pub struct SledStateStoreBuilderConfig {
 }
 
 impl SledStateStoreBuilder {
-    pub fn build(&mut self) -> Result<SledStateStore> {
-        let is_temp = self.path.is_none();
-
-        let mut cfg = Config::new().temporary(is_temp);
+    /// Path to the sled store files, created if not yet existing
```
Comment (jplatte) [inline crates/matrix-sdk-sled/src/state_store.rs:193]:
```suggestion
    /// Path to the sled store files, created if not it doesn't exist yet.
    ///
    /// Mutually exclusive with [`db`][Self::db], whichever is called last wins.
```

File: crates/matrix-sdk-sled/src/state_store.rs [hunk 6]
```diff
@@ -184,22 +190,40 @@ pub struct SledStateStoreBuilderConfig {
...
+    pub fn path(&mut self, path: PathBuf) {
+        self.db_or_path = Some(DbOrPath::Path(path));
+    }
 
-        let path = if let Some(path) = &self.path {
-            let path = path.join("matrix-sdk-state");
+    /// Use an existing open [`sled::Db`]
```
Comment (jplatte) [inline crates/matrix-sdk-sled/src/state_store.rs:198]:
```suggestion
    /// Use the given [`sled::Db`].
    ///
    /// Mutually exclusive with [`path`][Self::path], whichever is called last wins.
```

**Review by jplatte — APPROVED** [review]

**Comment by jplatte** [comment]
@gferon Could you have a look at the CI failures? The formatting is my fault, but there's also a documentation error and a test failure (looks like you forgot to return the builder from `path` and `db`).

File: crates/matrix-sdk-sled/src/state_store.rs [hunk 7]
```diff
@@ -184,22 +191,46 @@ pub struct SledStateStoreBuilderConfig {
...
+    pub fn path(&mut self, path: PathBuf) -> &mut SledStateStoreBuilder {
+        self.db_or_path = Some(DbOrPath::Path(path));
+        self
+    }
 
-        let path = if let Some(path) = &self.path {
-            let path = path.join("matrix-sdk-state");
+    /// Use the given [`sled::Db`].
+    ///
+    /// Mutually exclusive with [`path`][Self::path], whichever is called last wins.
```
Comment (jplatte) [inline crates/matrix-sdk-sled/src/state_store.rs:204]:
```suggestion
    /// Mutually exclusive with [`path`][Self::path], whichever is called last
    /// wins.
```

**Pull request merged**